ZENKO-3254: added two procedures to install and upgrade XDM (#1147)
scality/Zenko: docs/docsource/installation/install/installing_xdm.rst

.. _Installing XDM:

> Installing XDM

Reviewer: Install XDM (consistent with my previous comment)

==============

.. note::
   
   This installation procedure is performed on a single node.

Prerequisites
-------------

- Your disk must be partioned as follows:
   
  +-------------+--------+-----------+
  | Partition # |  Size  |  Service  |
  +=============+========+===========+

>   | 1           | 300GiB | Mongo     |

Reviewer 2: space between 300 and GiB (to apply everywhere)

  +-------------+--------+-----------+
  | 2           | 100GiB | Kafka     |
  +-------------+--------+-----------+
  | 3           | 100GiB | S3Data    |
  +-------------+--------+-----------+
  | 4           | 10GiB  | Redis     |
  +-------------+--------+-----------+
  | 5           | 10GiB  | Zookeeper |
  +-------------+--------+-----------+

  If necessary, partition your disk with:

  .. code::
      
     DISK_NAME=/dev/vdb
     parted -a none ${DISK_NAME} --script \
         mklabel gpt \
         mkpart primary ext4 1MiB 320GiB \
         mkpart primary ext4 320GiB 420GiB \
         mkpart primary ext4 420GiB 520GiB \
         mkpart primary ext4 520GiB 530GiB \
         mkpart primary ext4 530GiB 540GiB \
         mkpart primary ext4 540GiB 550GiB \
         mkpart primary ext4 550GiB 560GiB \
         mkpart primary ext4 560GiB 570GiB \
         mkpart primary ext4 570GiB 580GiB

Deploying MetalK8s
------------------

> For more information on MetalK8s installation, see *Installation* in the MetalK8s' documentation.

Reviewer 2: does not sound very good... Try to reword it :) 
You can avoid repeating MetalK8s and installation (hint: change "For more information about MetalK8s installation"). I'll let you think about it, if you cannot find anything let's talk about it

Reviewer: Install MetalK8s as described in its installation documentation.

Reviewer: If you want to italicize it, set it as "as described in *MetalK8s Installation*"

> Deploying XDM-Operator

Reviewer: That hyphen is up to no good. 

Deploy XDM Operator

----------------------

> #. From your terminal, create a namespace for XDM with:

Reviewer 2: too many  "with:"
From the terminal, create a namespace for XDM. Simpler and better :)

   .. code::

      /srv/scality/metalk8s-2.6.0/solutions.sh create-env --name zenko

#. Install XDM Base with:

   .. code::

      /srv/scality/metalk8s-2.6.0/solutions.sh import --archive $ZENKO_BASE_ISO
      sed "s/SOLUTION_ENV/zenko/g" /srv/scality/zenko-base-2.0.0-alpha.3/operator.      yaml | kubectl apply -f -
      kubectl -n zenko rollout status --timeout 10m deploy kubedb-operator
      sed "s/SOLUTION_ENV/zenko/g" /srv/scality/zenko-base-2.0.0-alpha.3/operator.      yaml | kubectl apply -f -

> #. Install XDM-Operator with:

Reviewer: Assuming "XDM Operator" is a thing... no hyphen, leave the O capitalized.

   .. code::

      /srv/scality/metalk8s-2.6.0/solutions.sh import --archive $ZENKO_ISO
      /srv/scality/metalk8s-2.6.0/solutions.sh activate --name zenko --version 2.0.      0-alpha.3
      /srv/scality/metalk8s-2.6.0/solutions.sh add-solution --name zenko       --solution zenko --version 2.0.0-alpha.3
      kubectl -n zenko rollout status --timeout 10m deploy zenko-operator

> Deploying XDM

Reviewer: Deploy XDM

          "attributes" : { },
          "authenticationFlowBindingOverrides" : { },
          "fullScopeAllowed" : true,
          "nodeReRegistrationTimeout" : -1,
          "protocolMappers" : [ {
            "name" : "instanceids_mapper",
            "protocol" : "openid-connect",
            "protocolMapper" : "oidc-usermodel-attribute-mapper",
            "consentRequired" : false,
            "config" : {
              "multivalued" : "true",
              "userinfo.token.claim" : "true",
              "user.attribute" : "instanceIds",
              "id.token.claim" : "true",
              "access.token.claim" : "true",
              "claim.name" : "instanceIds"
            }
          }, {
            "name" : "role_mapper",
            "protocol" : "openid-connect",
            "protocolMapper" : "oidc-usermodel-attribute-mapper",
            "consentRequired" : false,
            "config" : {
              "user.attribute" : "role",
              "id.token.claim" : "true",
              "access.token.claim" : "true",
              "claim.name" : "role",
              "userinfo.token.claim" : "true"
            }
          } ]
        } ]
      }
      EOF

> #. Create a yaml file for the new XDM's version with:

Reviewer: the new XDM version

      cat <<EOF | kubectl -n ringx-auth exec -i keycloak-0 -- /opt/jboss/keycloak/      bin/kcadm.sh create users -r "${REALM_NAME}" -f -   
      {
          "username": "${OIDC_USER}",
          "enabled": true,
          "totp": false,
          "emailVerified": true,
          "firstName": "zenko",
          "lastName": "tester",
          "email": "${OIDC_USER}@zenko.local",
          "attributes": {
              "instanceIds": [
                  "${INSTANCE_ID}"
              ],
              "role": [
              "user"
              ]
          },
          "credentials": [],
          "disableableCredentialTypes": [],
          "requiredActions": [],
          "realmRoles": [
              "uma_authorization",
              "offline_access"
          ],
          "clientRoles": {
              "account": [
              "view-profile",
              "manage-account"
              ]
          },
          "notBefore": 0,
          "groups": []
      }
      EOF
      
      kubectl -n ringx-auth exec -i keycloak-0 -- /opt/jboss/keycloak/bin/kcadm.sh       set-password -r ${REALM_NAME} --username ${OIDC_USER}

Testing
-------

> Via the User Interface

Reviewer: Using the GUI

**********************

> Add the following hosts to the browser user machine ``/etc/hosts`` to resolve to 

Reviewer: to your ``etc/hosts`` file, resolving them to the MetalK8s node's IP address:

MetalK8s node's IP address:

- management.zenko.local

- keycloak.zenok.local

- ui.zenko.local

- s3.zenko.local

.. note::

>    If the machine has port 80 open, you can access the User Interface through http://ui.zenko.local.

Reviewer: ```suggestion
   If port 80 is open, you can access the user interface through http://ui.zenko.local.
```

.. important::

>    The User Interface is limited to creating accounts and locations.

Reviewer: ```suggestion
   The user interface is limited to creating accounts and locations.
```

> Via Command Line

Reviewer: Using the Command Line

****************

Prerequisites
~~~~~~~~~~~~~~

- jq and curl must be installed

> - Access from the browser user machine

Reviewer: You must have browser access to the host

but... if you're doing this from the command line, why do you need browser access?

Author: I'm not sure of what that means, I asked @terrydervaux

              -H "Authorization: bearer $ACCESS_TOKEN" | \
              jq -rc '.instanceIds[0]'
      )

#. Create an account with:

   .. code::
      
      ZENKO_ACCOUNT='test-account-1'

      USER_PARAMS=$(
          echo '{}' |
          jq -c "
              .userName=\"${ZENKO_ACCOUNT}\" |
              .email=\"${ZENKO_ACCOUNT}@zenko.local\"
          "
      )
      
      curl -s -k -X POST \
          -H "X-Authentication-Token: ${TOKEN}" \
          -H "Content-Type: application/json" \
          -d "${USER_PARAMS}" \
          "http://management.zenko.local/api/v1/config/${INSTANCE_ID}/user" | \
          jq '.'

#. Create an additional S3 endpoint with:

   .. code::
      
      ENDPOINT_PARAMS=$(
          echo '{}' |
          jq -c "
              .hostname=\"${ENDPOINT_HOSTNAME}\" |
              .locationName=\"${LOCATION_NAME}\"
          "
      )
      
      curl -s -k -X POST \
          -H "X-Authentication-Token: ${TOKEN}" \
          -H "Content-Type: application/json" \
          -d "${ENDPOINT_PARAMS}" \
          "http://management.zenko.local/api/v1/config/${INSTANCE_ID}/endpoint" | \
          jq '.'

> #. Generate the account's key with:

Reviewer: "account key" is fine.

> - Your disk must be partioned as follows:

Reviewer: Partition your disk as follows:

   
  +-------------+---------+-----------+
  | Partition # | Size    | Service   |
  +=============+=========+===========+
  | 1           | 300 GiB | Mongo     |
  +-------------+---------+-----------+
  | 2           | 100 GiB | Kafka     |
  +-------------+---------+-----------+
  | 3           | 100 GiB | S3Data    |
  +-------------+---------+-----------+
  | 4           | 10 GiB  | Redis     |
  +-------------+---------+-----------+

>   | 5           | 10 GiB  | Zookeeper |

Reviewer: ZooKeeper
